Sources/EdgeNetworkHandlers/EdgeHitProcessor.swift

```diff
@@ -0,0 +1,74 @@
+//
+// Copyright 2020 Adobe. All rights reserved.
+// This file is licensed to you under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License. You may obtain a copy
+// of the License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under
+// the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+// OF ANY KIND, either express or implied. See the License for the specific language
+// governing permissions and limitations under the License.
+//
+
+import AEPServices
+import Foundation
+
+/// A `HitProcessing` which handles the processing of `EdgeHit`s
+class EdgeHitProcessor: HitProcessing {
+    private let LOG_TAG = "EdgeHitProcessor"
+    private var networkService: EdgeNetworkService
+    private var networkResponseHandler: NetworkResponseHandler
+    var retryInterval = TimeInterval(5)
+
+    init(networkService: EdgeNetworkService, networkResponseHandler: NetworkResponseHandler) {
+        self.networkService = networkService
+        self.networkResponseHandler = networkResponseHandler
+    }
+
+    func processHit(entity: DataEntity, completion: @escaping (Bool) -> Void) {
+        guard let data = entity.data, let edgeHit = try? JSONDecoder().decode(EdgeHit.self, from: data) else {
+            // can't convert data to hit, unrecoverable error, move to next hit
+            Log.debug(label: LOG_TAG, "processHit - Failed to decode edge hit '\(entity.uniqueIdentifier)'.")
+            completion(true)
+            return
+        }
+
+        let callback: ResponseCallback = NetworkResponseCallback(requestId: entity.uniqueIdentifier, responseHandler: networkResponseHandler)
+        let hitCallback = EdgeHitResponseCallback(completion: completion, callback: callback)
```

Comment:
I wish we could simplify it, but we can't (I haven't found a way). This is a direct product of using a protocol based callback pattern instead of a traditional callback pattern seen through other Swift APIs. The real fix here is to move away from the protocol based callbacks and use completion handlers.

Reply:
As we discussed, you might need to refactor the doRequest logic a bit to cover the retries until success/unrecoverable error. Additionally when there is no network connection we need to wait for a longer period of time before retrying. 
This might also allow you to remove the double callback in here.